server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1507,6 +1509,85 @@ public void testSnapshotFromStreamInput() throws IOException {
+        List<Translog.Index> operations = new ArrayList<>();
+        int translogOperations = randomIntBetween(10, 100);
+        int maxTrimmedSeqNo = translogOperations - randomIntBetween(4, 8);
+
+        for (int op = 0; op < translogOperations; op++) {
+            String ascii = randomAlphaOfLengthBetween(1, 50);
+            Translog.Index operation = new Translog.Index("test", "" + op, op,
+                primaryTerm.get(), ascii.getBytes("UTF-8"));
+            operations.add(operation);
+        }
+        // shuffle a bit - move several first items to the end
+        for(int i = 0, len = randomIntBetween(5, 10); i < len; i++){
+            operations.add(operations.remove(0));
+        }
+
+        for (Translog.Index operation : operations) {
+            translog.add(operation);
+        }
+
+        translog.rollGeneration();
+
+        translog.trim(primaryTerm.get() + 1, maxTrimmedSeqNo);
+        translog.sync();
+
+        Set<Long> expectedSeqNo = LongStream.range(0, maxTrimmedSeqNo + 1).boxed().collect(Collectors.toSet());
+        try (Translog.Snapshot snapshot = translog.newSnapshot()) {
+            Translog.Operation next;
+            while ((next = snapshot.next()) != null) {
+                assertThat(expectedSeqNo.remove(next.seqNo()), is(true));
+            }
+            assertThat(snapshot.skippedOperations(), is(translogOperations - (maxTrimmedSeqNo + 1)));
+        }
+        assertThat(expectedSeqNo, hasSize(0));
+    }
+
+    public void testExceptionOnTrimAboveSeqNo( ) throws Exception {
+        Path tempDir = createTempDir();
+        final FailSwitch fail = new FailSwitch();
+        fail.failNever();
+        TranslogConfig config = getTranslogConfig(tempDir);
+        final Translog failableTLog = getFailableTranslog(fail, config, randomBoolean(), false, null, createTranslogDeletionPolicy());
+
+        List<Translog.Index> operations = new ArrayList<>();
+        int translogOperations = randomIntBetween(10, 100);
+        int maxTrimmedSeqNo = translogOperations - randomIntBetween(4, 8);
+
+        for (int op = 0; op < translogOperations; op++) {
+            String ascii = randomAlphaOfLengthBetween(1, 50);
+            Translog.Index operation = new Translog.Index("test", "" + op, op,
+                primaryTerm.get(), ascii.getBytes("UTF-8"));
+            operations.add(operation);
+        }
+        // shuffle a bit - move several first items to the end
+        for(int i = 0, len = randomIntBetween(5, 10); i < len; i++){
+            operations.add(operations.remove(0));
+        }
+
+        for (Translog.Index operation : operations) {
+            failableTLog.add(operation);
+        }
+
+        failableTLog.rollGeneration();
+        fail.failAlways();
+        try {
+            failableTLog.trim(primaryTerm.get() + 1, maxTrimmedSeqNo);
+            fail();
+        } catch (MockDirectoryWrapper.FakeIOException ex) {
+            // all is fine
+        }
+
```

Comment:
please make sure all files closed and no file is leaked.

Reply:
do we really need to manually check that all files are closed while that check is provided by ESTestCase ?

Reply:
If that's the case, we don't need - that's making sure :D - where do you see it's done in ESTestCase? (I didn't check myself)

Reply:
*TestRuleTemporaryFilesCleanup* takes care of it (in case I don't do _close_ of file handlers):

```
java.lang.RuntimeException: file handle leaks: [FileChannel(/private/var/folders/dz/750jvg8j31j21s52xyhzcj4c0000gn/T/org.elasticsearch.index.translog.TranslogTests_FBDFE9B843A061D2-001/tempDir-003/translog-1.tlog), FileChannel(/private/var/folders/dz/750jvg8j31j21s52xyhzcj4c0000gn/T/org.elasticsearch.index.translog.TranslogTests_FBDFE9B843A061D2-001/tempDir-003/translog-2.tlog), FileChannel(/private/var/folders/dz/750jvg8j31j21s52xyhzcj4c0000gn/T/org.elasticsearch.index.translog.TranslogTests_FBDFE9B843A061D2-001/tempDir-003/translog-3.tlog)]

	at __randomizedtesting.SeedInfo.seed([FBDFE9B843A061D2]:0)
	at org.apache.lucene.mockfile.LeakFS.onClose(LeakFS.java:63)
	at org.apache.lucene.mockfile.FilterFileSystem.close(FilterFileSystem.java:77)
	at org.apache.lucene.mockfile.FilterFileSystem.close(FilterFileSystem.java:78)
	at org.apache.lucene.util.TestRuleTemporaryFilesCleanup.afterAlways(TestRuleTemporaryFilesCleanup.java:228)
	at com.carrotsearch.randomizedtesting.rules.TestRuleAdapter$1.afterAlways(TestRuleAdapter.java:31)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:43)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.Exception
	at org.apache.lucene.mockfile.LeakFS.onOpen(LeakFS.java:46)
	at org.apache.lucene.mockfile.HandleTrackingFS.callOpenHook(HandleTrackingFS.java:81)
	at org.apache.lucene.mockfile.HandleTrackingFS.newFileChannel(HandleTrackingFS.java:197)
	at org.apache.lucene.mockfile.HandleTrackingFS.newFileChannel(HandleTrackingFS.java:166)
	at java.base/java.nio.channels.FileChannel.open(FileChannel.java:292)
	at java.base/java.nio.channels.FileChannel.open(FileChannel.java:340)
	at org.elasticsearch.index.translog.Translog.openReader(Translog.java:279)
	at org.elasticsearch.index.translog.Translog.recoverFromFiles(Translog.java:225)
	at org.elasticsearch.index.translog.Translog.<init>(Translog.java:177)
	at org.elasticsearch.index.translog.TranslogTests$4.<init>(TranslogTests.java:2128)
	at org.elasticsearch.index.translog.TranslogTests.getFailableTranslog(TranslogTests.java:2128)
	at org.elasticsearch.index.translog.TranslogTests.testExceptionOnTrimAboveSeqNo(TranslogTests.java:1552)
```